feat: high-availability validator signer

spypsy opened the pull request:
Fixes [A-352](https://linear.app/aztec-labs/issue/A-352/add-postgresql-integration-for-recording-signed-duties) [A-353](https://linear.app/aztec-labs/issue/A-353/implement-core-slash-protection-logic)

Introduces High-Availability validator signer. The signer uses a Postgres DB that will be used by multiple sequencer nodes, running with the same set of validator keys, in order to ensure that only one payload per duty is signed.

PhilWindle commented on yarn-project/validator-ha-signer/src/db/postgres.ts:
```diff
@@ -0,0 +1,204 @@
+  /**
+   * Find an existing duty record
+   */
+  async findDuty(validatorAddress: EthAddress, slot: bigint, dutyType: DutyType): Promise<ValidatorDutyRecord | null> {
```
Why does return `null`?

PhilWindle commented:
I feel like the database interactions are convoluted and have potential race conditions. IIUC we are manually performing quite a complicated check and set operation across multiple calls. Can this not be reduced to about 3 calls and have the DB provide the concurrency guarantees:

1. Insert record if not present, return record if present, We repeatedly call this if record is present and not yet signed.
2. Update signed record. Called upon success.
3. Delete failed record. Called upon failure.

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -0,0 +1,380 @@
+
+      // Should release client
+      expect(mockClient.release).toHaveBeenCalled();
+    });
```
I think this test is too brittle considering how tightly coupled it is to the migration being run.

alexghr commented on yarn-project/validator-ha-signer/migrations/1_initial-schema.ts:
I see there's a `src/db/schema.ts` file too which defines tables and constraints. Are both files needed?

spypsy replied:
The `SCHEMA_SETUP in `schema.ts` could be removed I guess, and use what's in the migration code to initialize. Would that make sense?

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -0,0 +1,380 @@
+
+    // Create mock pool
+    mockPool = mock<Pool>();
+    mockPool.connect.mockImplementation(() => Promise.resolve(mockClient));
```
Instead of using a mocked Posgres connection, we could instantiate [pglite](https://github.com/electric-sql/pglite) (Postgres compiled to wasm) in a temp folder. We could eliminate all mocked calls this way.

spypsy replied:
yep, was just looking into some form of postgres to use for proper testing bc this mock doesn't provide much confidence. Will check `pglite` 👍

alexghr commented on yarn-project/validator-ha-signer/src/db/schema.ts:
```diff
@@ -0,0 +1,202 @@
+
+/**
+ * Query to clean up old duties (for maintenance)
+ * Removes duties older than a specified timestamp
+ */
+export const CLEANUP_OLD_DUTIES = `
+DELETE FROM validator_duties
+WHERE started_at < $1;
+`;
```
I don't think this query is using the `idx_validator_duties_status` index since it's defined on both `status` and `started_at` while the query does not specify status so it might require a full table scan.

alexghr replied:
Should this only clean 'completed' duties?

spypsy replied:
was just added to cleanup **all** duties if needed. Can add one for `signed` duties only.
Would the way to make this one more efficient be:
```
DELETE FROM validator_duties
WHERE status IN ('signing', 'signed', 'failed')
  AND started_at < $1;
```
?

alexghr commented on yarn-project/validator-ha-signer/src/db/schema.ts:
I like keeping the SQL queries separate but I'm slightly worried that if we change the query here and forget to update their uses in `PostgresSlashingProtectionDatabase` we'll only know of the issue at runtime since the unit tests use mocked clients that don't actually check the query syntax.

If we added `pglite` then we could run the queries in unit tests.

alexghr commented on yarn-project/validator-ha-signer/src/db/types.ts:
```diff
@@ -0,0 +1,90 @@
+  /** Ethereum address of the validator */
+  validatorAddress: EthAddress;
+  /** Slot number for this duty */
+  slot: bigint;
```
These should probably use the new `SlotNumber`, `BlockNumber` tagged number types.

spypsy replied:
yep, this is done in follow-up PR: https://github.com/AztecProtocol/aztec-packages/pull/19094/changes#diff-9bdcb259e1719481ea7fd0cc3cd47199d34833abb0a207f2eea1ea8f45639873R31

alexghr commented on yarn-project/validator-ha-signer/src/config.ts:
```diff
@@ -0,0 +1,120 @@
+  nodeId: {
+    env: 'SLASHING_PROTECTION_NODE_ID',
+    description: 'The unique identifier for this node',
+    ...secretStringConfigHelper(`high-availability-node-${Math.random().toString(36).substring(2, 6).toUpperCase()}`),
```
Two questions here:
1. does this need to be secret?
2. the config type defines this field as `string` but the parser returns `SecretValue<string>`.

alexghr replied:
Two more questions:
1. If an operator does not set this value in their node, so it's chosen randomly and the validator crashes, will it find and update its own duties the next time it comes back up?
2. What happens if as an operator I do set the node-id but.. I use the same value for two different nodes?

spypsy replied:
on the first questions, this was a mistaken import and I also fixed here https://github.com/AztecProtocol/aztec-packages/pull/19094/changes#diff-6f4576f0c6c73e00ee9aaee175898098f299879c2eef74a5ee407118dc9e7998R56
I noticed the issue when testing and I was using the follow-up branch so fixed there 😬 

As for the follow-up questions, `nodeId` is just metadata atm but q1 does bring up a good point on what happens if a validator crashes during `SIGNING` (even if it recovers with the same correct `nodeId`). 
Basically all nodes (including the one that failed) will keep seeing `signing` as the status until they timeout and return `DutyAlreadySignedError(slot, dutyType, 'unknown (timeout)')`. 

This is actually in-line with the agreed SLAs where we stated that if a node fails while signing its duty, we can't guarantee that the HA-setup will recover (for that particular duty). 
So should definitely be fixed but tempted to make a new ticket for it as it's non-trivial.

Probably one simple improvement to make for now is to **not** auto-generate an ID so the user **has** to pick one

PhilWindle replied:
I agree, I don't think we should wuto generate an Id.

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+
+    this.log.debug(`Checking duty: ${dutyType} for slot ${slot}`, {
+      validatorAddress: validatorAddress.toString(),
+      nodeId,
```
This is going to show `nodeId: "<Redacted>"`

spypsy replied:
fixed in #19094

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+      // Handle existing duty - may throw or delete failed records
+      await this.handleExistingDuty(existingDuty, params);
+      // If we reach here, the failed duty was deleted, proceed with insertion
+    }
```
This is a race condition. The node needs to always attempt insertion and handle insertion success/failure appropriately

spypsy replied:
most of this has now been updated, going through it to check what's still relevant

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+    let inserted = await this.db.insertDuty(params);
+
+    // Loop until we acquire the lock or get a definitive result
+    while (!inserted) {
```
are there any situations under which this `while` loop never exits?

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+            // Record disappeared (shouldn't happen, but handle gracefully)
+            // Signal caller to try acquiring the lock
+            return WaitResult.READY_TO_ACQUIRE_LOCK;
+          }
```
This is also a race condition. By the time the code reaches `this.db.insertDuty` in the `while` loop in `checkAndRecord` some other node may have acquired the lock.

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+              validatorAddress: validatorAddress.toString(),
+              failedNode: duty.nodeId,
+            });
+            await this.db.deleteFailedDuty(validatorAddress, slot, dutyType);
```
IIRC the spec said we wouldn't retry failed slots? I think deleting a lock could be error prone.

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+   * Either throws an error (duty already signed) or returns to allow insertion (failed duty deleted).
+   * For in-progress duties, waits and then either throws or returns.
+   */
+  private async handleExistingDuty(existingDuty: ValidatorDutyRecord, params: CheckAndRecordParams): Promise<void> {
```
I'm not a big fan of using `throw` for control flow. I think it would be clearer if this function returned different statuses for the parent function to handle

spypsy replied:
I tend to agree but this is a lot simpler to bubble up all the way to building blocks / signing attestations in validator application logic, rather than adding a bunch more if / else handlers all the way up

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,275 @@
+      validatorAddress: validatorAddress.toString(),
+      nodeId,
+    });
+  }
```
Is there any protection against rogue writes? E.g. some bug in the validator that causes NODE-FOO to call `recordSuccess`/`recordFailure` when the lock is actually held by NODE-BAR?

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.test.ts:
```diff
@@ -0,0 +1,391 @@
+      const params2 = { ...params, nodeId: NODE_ID_2 };
+      const promise = service.checkAndRecord(params2);
+
+      // Complete first node's signing after a short delay
+      await sleep(100);
+      await service.recordSuccess({
```
what would happen if the node never completes the signature? Does node 2 recover?

spypsy replied:
nope, as mentioned in another comment if a node fails during 'signing' we can't currently recover.

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,144 @@
+   */
+  async signWithProtection(
+    validatorAddress: EthAddress,
+    signingRoot: Buffer32,
```
Nit: I'd rename `signingRoot` to just `message` or `messageHash` to match the rest of the codebase.

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,144 @@
+      throw error;
+    }
+
+    // Record success
+    await this.slashingProtection.recordSuccess({
+      validatorAddress,
+      slot,
+      dutyType,
+      signature,
+      nodeId: this.config.nodeId,
+    });
```
This doesn't get retried so I think we'll lose signature in case of temporary network issues.

alexghr commented on yarn-project/validator-ha-signer/migrate.ts:
This is the third(?) migration script

PhilWindle commented on yarn-project/validator-ha-signer/src/db/schema.ts:
```diff
@@ -0,0 +1,202 @@
+/**
+ * Query to update a duty to 'signed' status
+ */
+export const UPDATE_DUTY_SIGNED = `
```
These UPDATE/DELETE queries should require the node to that which made the insert shouldn't they?

`WHERE node_id = $5` or something.

spypsy replied:
updated to use lock_token

PhilWindle commented on yarn-project/validator-ha-signer/MIGRATIONS.md:
```diff
@@ -0,0 +1,192 @@
+    spec:
+      initContainers:
+        - name: db-migrate
+          image: validator-image:latest
```
What's 'validator-image'?

spypsy replied:
sry didn't realize claude added this guide when I first created the migrations code.
Seems kinda useful, I've updated to correctly reflect how it currently works (which is with `aztec migrate-ha-db` command)

PhilWindle commented on yarn-project/validator-ha-signer/README.md:
```diff
@@ -0,0 +1,155 @@
+    spec:
+      containers:
+        - name: migrate
+          image: your-validator-image
```
Should this be `aztecprotocol/aztec`?

PhilWindle replied:
And wouldn't it use the CLI command?

spypsy replied:
Updated REAME, it wasn't up-to-date

PhilWindle commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,232 @@
+   * @returns true if the update succeeded, false if token didn't match
+   */
+  async recordFailure(params: RecordFailureParams): Promise<boolean> {
+    const { validatorAddress, slot, dutyType, error, lockToken } = params;
```
This still feels redundant. Can't we simply delete the record?

spypsy replied:
hmm yeah I think you're right. other than logging/observability this is a better solution as it simplifies things even more.

PhilWindle commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,232 @@
+
+    while (true) {
+      // insert if not present, get existing if present
+      const { isNew, record } = await this.db.tryInsertOrGetExisting(params);
```
If node A acquires the lock and then dies, how long is node B going to keep looping? Is it infinite? At most it should be until the end of the slot, but I can't see where it would time out.

spypsy replied:
Further down it does:
```
 } else if (record.status === DutyStatus.SIGNING) {
    // Another node is currently signing - check for timeout
    if (Date.now() - startTime > this.signingTimeoutMs) {
      this.log.warn(`Timeout waiting for signing to complete for duty ${dutyType} at slot ${slot}`, {
        validatorAddress: validatorAddress.toString(),
        timeoutMs: this.signingTimeoutMs,
        signingNodeId: record.nodeId,
      });
      throw new DutyAlreadySignedError(slot, dutyType, 'unknown (timeout)');
    }

  // .... keep looping
}
```
Could  switch from having a fixed timeout to having a deadline which would be the end of the slot.
Can implement that in the other PR

alexghr commented on yarn-project/validator-ha-signer/src/db/migrations/1_initial-schema.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  // Insert initial schema version
+  pgm.sql(`
+    INSERT INTO schema_version (version)
+    VALUES (${SCHEMA_VERSION})
+    ON CONFLICT (version) DO NOTHING;
+  `);
+}
```
Doesn't pg-migrate already create a schema migration table?

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -0,0 +1,590 @@
+    await db.close();
+  });
+
+  describe('schema setup', () => {
```
I think we can delete these tests as they're just testing PGlite functions correctly (which shouldn't be our concern)

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -0,0 +1,590 @@
+    it('should handle large block numbers correctly', async () => {
+      const largeBlockNumber = 9007199254740991n;
+
+      const result = await db.query<InsertOrGetRow>(INSERT_OR_GET_DUTY, [
```
I think these queries should be tested through our `PostgresSlashingProtectionDatabase` implementation rather than going through the database. Doing it this way would also validate that the positional query parameters in the service are passed correctly

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.ts:
```diff
@@ -0,0 +1,202 @@
+    const row = result.rows[0];
+    return {
+      isNew: row.is_new,
+      record: this.rowToRecord(row),
```
does this return the lock token?

alexghr replied:
It's worth adding a comment here specifying that on failing to get the lock the returned 'lock_token' from the db is going to be the empty string.

spypsy replied:
I did initially return the lock_token for non-new records. It shouldn't be an issue since this is a trusted setup but also wouldn't that defeat its purpose?

alexghr replied:
Oh I meant in the 'failed lock' case. It shouldn't return the lock token to Node B if Node A has the lock and the SQL handles this correctly.

alexghr commented on yarn-project/validator-ha-signer/src/db/schema.ts:
```diff
@@ -0,0 +1,236 @@
+    error_message,
+    TRUE as is_new
+)
+SELECT * FROM inserted
+UNION ALL
+SELECT
```
This looks quite complicated but I get how it's meant to work.

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -0,0 +1,590 @@
+    for (const statement of SCHEMA_SETUP) {
+      await db.query(statement);
+    }
+    await db.query(INSERT_SCHEMA_VERSION, [SCHEMA_VERSION]);
```
Does this duplicate `setupTestSchema`?

spypsy replied:
yep..

alexghr commented on yarn-project/validator-ha-signer/src/factory.test.ts:
```diff
@@ -0,0 +1,91 @@
+
+    // Cast to PostgresSlashingProtectionDatabase to access close method
+    await (db as PostgresSlashingProtectionDatabase).close();
+    expect(mockPool.end).toHaveBeenCalled();
```
Shouldn't this test be moved to the `postgres.test.ts` file?

alexghr commented on yarn-project/validator-ha-signer/src/factory.test.ts:
is this file only meant to test config?

spypsy replied:
yeah feels a bit pointless tbh. Will remove these and add  a test with pglite to see that it runs migrations correctly when `runMigrations` option is true

spypsy replied:
aah can't run migrations on pglite (node-pg-migrate doesn't support it).
Will just go ahead as we discussed and not have any `runMigrations` option and can remove this test

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.test.ts:
```diff
@@ -0,0 +1,522 @@
+      }
+    });
+  });
+
```
I'd like to see a test where two nodes try to acquire a lock _at the same time_, i.e. something like `Promise.all([service.checkAndRecord({...params, nodeId: 'a'}), service.checkAndRecord({...params, nodeId: 'b'})])`

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.test.ts:
```diff
@@ -0,0 +1,522 @@
+
+      expect(success).toBe(false);
+      // Duty should still be in signing state
+      const result = await db.tryInsertOrGetExisting(params);
+      expect(result.record.status).toBe(DutyStatus.SIGNING);
```
This is an upsert so this doesn't actually assert that the insertion was done correctly. This function could have just inserted the row and this test wouldn't know.

alexghr replied:
NVM, I saw the tests further down in this test suite.

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.test.ts:
```diff
@@ -0,0 +1,522 @@
+
+      // First one should succeed, let it complete signing
+      await sleep(50);
+      const result = await db.tryInsertOrGetExisting(params1);
```
This makes node1 the winner. We can't assume the first promise will be the one to win. This should use a `Promise.race`

spypsy replied:
now I added a `Promise.allSettled` as per other comment I think this one becomes redundant

spypsy replied:
kept just one test where 3 nodes all fire at once with Promise.race and we check expected results

alexghr commented on yarn-project/validator-ha-signer/src/slashing_protection_service.ts:
```diff
@@ -0,0 +1,214 @@
+          throw new SlashingProtectionError(slot, dutyType, record.messageHash, messageHash);
+        }
+        throw new DutyAlreadySignedError(slot, dutyType, record.nodeId);
+      } else if (record.status === DutyStatus.SIGNING) {
```
Are there any other statuses to check? "error" comes to mind

alexghr replied:
Ah error was removed! Can we add a 
```
else {
  const _: never
}
```
or throw an exception in case a new status gets added?

spypsy replied:
feels like it should error for unknown status, adding

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,164 @@
+  private readonly slashingProtection: SlashingProtectionService | undefined;
+
+  constructor(
+    db: SlashingProtectionDatabase | undefined,
```
Shouldn't this be defined? Otherwise I'd expect the node to use a different Signer implementation.

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,164 @@
+    } else {
+      this.slashingProtection = undefined;
+      this.log.info('Validator HA Signer initialized WITHOUT slashing protection');
+    }
```
this is dangerous: I want HA but I have no slashing protection. We should throw in this case IMHO

alexghr reviewed (commented):
Looks good but I'd like to see the HASigner check for invalid configs. The test changes would be nice to have though.

PhilWindle commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
````diff
@@ -0,0 +1,164 @@
+ * );
+ * ```
+ */
+export class ValidatorHASigner {
````
I think this class shold be changed to only support fully protected signing, no fall through. Then we want a fatory to create the appropriate type at runtime.

spypsy reviewed (commented):
responses

spypsy committed:
feat: high-availability validator signer

Fixes [A-352](https://linear.app/aztec-labs/issue/A-352/add-postgresql-integration-for-recording-signed-duties) [A-353](https://linear.app/aztec-labs/issue/A-353/implement-core-slash-protection-logic)

Introduces High-Availability validator signer. The signer uses a Postgres DB that will be used by multiple sequencer nodes, running with the same set of validator keys, in order to ensure that only one payload per duty is signed.

PhilWindle commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,164 @@
+ */
+export class ValidatorHASigner {
+  private readonly log: Logger;
+  private readonly slashingProtection: SlashingProtectionService | undefined;
```
This can't be undefined can it?

spypsy replied:
I've updated this in new PR FYI

PhilWindle commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -0,0 +1,164 @@
+  ): Promise<Signature> {
+    // If slashing protection is disabled, just sign directly
+    if (!this.slashingProtection) {
+      this.log.info('Signing without slashing protection enabled', {
```
I don't think this branch is possible.

spypsy replied:
same here, removed
